From 81ebd97083a4b0a7be8eba4f2611c27ad4a956cf Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Wed, 13 Jan 2021 01:40:38 +0200
Subject: [PATCH] [SW-0] preboot: fix linker symbols usage

The linker symbols must be used with a reference to them, never the
variables themselves.
Fix the linker symbols usage and also the extern declaration.

Change-Id: I21e5eaed9f2e344e8f3a9fb58c81c92550b44aac
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/include/preboot.h b/include/preboot.h
index da01ae2..d674bda 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -26,6 +26,7 @@
 #ifndef __ASSEMBLY__
 
 #include <stddef.h>
+#include <stdint.h>
 #include <stdbool.h>
 #include <types.h>
 #include "auto_generated.h"
@@ -60,8 +61,8 @@
 	HL_DEBUG_HW_DBG_CARD,
 };
 
-extern void *_start;		/* Start of ppboot/preboot region */
-extern void *__stack_top;	/* End of ppboot/preboot region */
+extern uintptr_t _start;	/* Start of ppboot/preboot region */
+extern uintptr_t __stack_top;	/* End of ppboot/preboot region */
 
 static inline u8 regmap_readb(u32 offset)
 {
diff --git a/preboot/preboot.c b/preboot/preboot.c
index ef5b81c..568785c 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -212,9 +212,7 @@
 
 static bool preboot_is_overlap_address_range(u64 addr, int len)
 {
-	void *addrptr = (void *)(uintptr_t)addr;
-
-	if (addrptr + len <= _start || addrptr > __stack_top)
+	if (addr + len <= (uintptr_t)&_start || addr >= (uintptr_t)&__stack_top)
 		return false;
 
 	return true;
